src/main/scala/com/fulcrumgenomics/testing/VcfBuilder.scala

```diff
@@ -0,0 +1,232 @@
+      attrs ++
+        (if (dp < 0) Map.empty else Map("DP" -> dp)) ++
+        (if (ad.isEmpty) Map.empty else Map("AD" -> ad.toIndexedSeq)) ++
+        (if (pl.isEmpty) Map.empty else Map("DP" -> pl.toIndexedSeq))
+    }
+  }
+
+  /** Constructs a VcfBuilder using the supplied header. */
+  def apply(header: VcfHeader): VcfBuilder = new VcfBuilder(header)
+
+  /** Constructs a VcfBuilder using the [[DefaultHeader]] and the set of samples provided. */
+  def apply(samples: Seq[String]): VcfBuilder = {
+    require(samples.distinct.size == samples.size, s"${samples.mkString(",")} contains duplicate sample names.")
+    new VcfBuilder(DefaultHeader.copy(samples=samples.toIndexedSeq))
+  }
+}
+
+/** Class for building VCFs for testing purposes. */
+class VcfBuilder private (initialHeader: VcfHeader) extends Iterable[Variant] {
+  /** Genomic location as (sequence_index, position). */
+  private type Location = (Int, Int)
+  private var _header: VcfHeader = initialHeader
+  private val variants: mutable.Map[Location, Variant] = mutable.TreeMap()
+
+  /** Provides access to the header that will be written to the VCF. */
+  def header: VcfHeader = _header
+
+  /** Adds one or more INFO headers to the VCF Header. */
+  def withInfoHeaders(headers: VcfInfoHeader*): this.type = {
+    this._header = this._header.copy(infos = this._header.infos ++ headers)
+    this
+  }
+
+  /** Adds one or more FORMAT headers to the VCF Header. */
+  def withFormatHeaders(headers: VcfFormatHeader*): this.type = {
+    this._header = this._header.copy(formats = this._header.formats ++ headers)
+    this
+  }
+
+  /** Adds one or more FILTER headers to the VCF Header. */
+  def withFilterHeaders(headers: VcfFilterHeader*): this.type = {
+    this._header = this._header.copy(filters = this._header.filters ++ headers)
+    this
+  }
+
+  /** Adds one or more non-INFO/FORMAT/FILTER headers to the VCF Header. */
+  def withOtherHeaders(headers: VcfGeneralHeader*): this.type = {
+    this._header = this._header.copy(others = this._header.others++ headers)
+    this
+  }
+
+  /** Adds a variant to the set being built.  The variant should contain all information required as it is not
+    * possible to update a variant once added. If a variant already exists at the given position an exception
+    * will be thrown.
+    *
+    * The genotypes are specified using instance of the [[Gt]] helper class.  The genotype strings within the [[Gt]]
+    * objects may be either numeric like in a VCF (e.g. `0/1`) or use allele strings (e.g. `A/T`.)
+    *
+    * The variant must also be valid, e.g. not reference INFO, FILTER or FORMAT fields that are no in the header,
+    * and not have alleles in genotypes that are not in the set of alleles for the variant.
+    *
+    * If the header contains more samples than there are genotype give, simple diploid no-call genotypes will
+    * be inserted for the remaining samples.
+    */
+  def add(chrom: String = this.header.contigs.head.name,
```

Comment:
I feel like you have a lot of comments on `VcfBuilder` based on your suggestion that it should be for more than just testing.  Right now, this is intended solely for testing, based on it being in the `testing` package.  If we want to make it robust enough to handle all user inputs, and have it be for more testing we can do that, but I think that's a separate PR because that's a significant amount of work.